kedro/contrib/io/pyspark/spark_hive_data_set.py

```diff
@@ -0,0 +1,304 @@
+                    database=self._database, table=self._table
+                )
+            )
+        return self._get_spark().sql(
+            "select * from {database}.{table}".format(
+                database=self._database, table=self._table
+            )
+        )
+
+    def _save(self, data: DataFrame) -> None:
+        if not self._exists():
+            self._create_empty_hive_table(data)
+            self._table_columns = data.columns
+        self._validate_save(data)
+        write_methods = {
+            "insert": self._insert_save,
+            "upsert": self._upsert_save,
+            "overwrite": self._overwrite_save,
+        }
+        write_methods[self._write_mode](data)
+
+    def _insert_save(self, data: DataFrame) -> None:
+        data.createOrReplaceTempView("tmp")
+        self._get_spark().sql(
+            "insert into {database}.{table} select {columns} from tmp".format(
+                database=self._database,
+                table=self._table,
+                columns=", ".join(self._table_columns),
+            )
+        )
+
+    def _upsert_save(self, data: DataFrame) -> None:
+        if self._load().rdd.isEmpty():
+            self._insert_save(data)
+        else:
+            joined_data = data.alias("new").join(
+                self._load().alias("old"), self._table_pk, "outer"
+            )
+            upsert_dataset = joined_data.select(
+                [  # type: ignore
+                    coalesce(
+                        "new.{}".format(col_name), "old.{}".format(col_name)
+                    ).alias(col_name)
+                    for col_name in set(data.columns)
+                    - set(self._table_pk)  # type: ignore
+                ]
+                + self._table_pk
+            )
+            temporary_persisted_tbl_name = "temp_{}".format(uuid.uuid4().int)
+            with StagedHiveDataSet(
+                upsert_dataset,
+                stage_database_name=self._database,
+                stage_table_name=temporary_persisted_tbl_name,
+            ) as temp_table:
+                self._overwrite_save(temp_table.staged_data)
+
+    def _overwrite_save(self, data: DataFrame) -> None:
+        self._get_spark().sql(
+            "truncate table {database}.{table}".format(
+                database=self._database, table=self._table
+            )
+        )
+        self._insert_save(data)
+
+    def _validate_save(self, data: DataFrame):
+        hive_dtypes = set(self._load().dtypes)
+        data_dtypes = set(data.dtypes)
+        if data_dtypes != hive_dtypes:
+            new_cols = data_dtypes - hive_dtypes
+            missing_cols = hive_dtypes - data_dtypes
+            raise DataSetError(
+                "dataset does not match hive table schema.\n"
+                "Present on insert only: {new_cols}\n"
+                "Present on schema only: {missing_cols}".format(
+                    new_cols=new_cols, missing_cols=missing_cols
```

Comment:
Maybe worth sorting them. Otherwise test checks may fail as the order is not guaranteed for sets.